RFCs/FS-1333-record-union-constructor-accessibility.md

```diff
@@ -0,0 +1,401 @@
+
+In contrast, the following constructs are _not_ subject to this accessibility:
+
+* Record field read and (if mutable) write access:
+
+  ```fsharp
+  type R =
+    { x: int
+      mutable y: int }
+    private new
+    static member Default = { x = 1; y = 2 }
+
+  let r = R.Default
+  r.x      // OK
+  r.y <- 3 // OK
+  ```
+
+* Record deconstruction:
+
+  ```fsharp
+  type R =
+    { x: int }
+    private new
+    static member Default = { x = 1 }
+
+  match R.Default with
+  | { x = 2 } -> "two" // OK
+  | _ -> "something else"
+  // val it: string = "something else"
+  ```
+
+* Anonymous record copy-and-update syntax using the given type as source:
+
+  ```fsharp
+  type R =
+    { x: int }
+    private new
+    static member Default = { x = 1 }
+
+  let r = {| R.Default with y = "a" |} // OK
+  // val r: {| x: int; y: string |} = { x = 1; y = "a" }
+  ```
+
+* Union deconstruction:
+
+  ```fsharp
+  type U =
+    | Case1 of int
+    | Case2
+    private new
+
+  let f (u: U) =
+    match u with
+    | Case1 x -> x // OK
+    | Case2 -> 0   // OK
+  // val f: u -> int = <fun>
+  ```
+
+* `IsX` case check properties:
+
+  ```fsharp
+  type U =
+    | Case1 of int
+    | Case2
+    private new
+
+  let f (u: U) =
+    if u.IsCase1 then 1 else 2 // OK
+  // val f: u -> int = <fun>
+  ```
+
+* `Tag` property, nested `Tags` enum and nested classes per case from C#:
+
+  ```fsharp
+  type U =
+    | Case1 of int
+    | Case2
+    private new
+    static member Default = Case1 1
+  ```
+
+  ```csharp
+  var x = U.Default.Tag == (int)U.Tags.Case1 ? (x as U.Case1).Item : 0; // OK
+  ```
+
+## Accessibility hierarchy
+
+If the representation accessibility is specified and the constructor accessibility is not,
+then the representation accessibility applies to the construction constructs listed above.
+This is unchanged from the existing behaviour.
+
+```fsharp
+type R =
+  private { x: int }
+
+let r = { x = 1 } // error FS1093: The union cases or fields of the type 'R' are not accessible from this code location
+```
+
+If both the representation accessibility and the constructor accessibility are specified,
+then the more restrictive of the two applies to the construction constructs listed above,
+and the representation accessibility applies to other representation constructs such as field access and deconstruction.
+
+**Alternate design**: Should the constructor accessibility be applied instead, even if it is less restrictive than the representation accessibility?
+It does seem strange to be able to construct a value using field names, and then to be unable to access said fields.
+Reasonable use cases would be welcome in support of this variation.
+
+**Design question**: If we do keep the design as specified here, should there be a warning, or even an error,
+when the constructor accessibility is less restrictive than the representation accessibility?
+The constructor accessibility will essentially be ignored.
```

Comment:
IL allows that, but I would tend towards a warning if this scenario is detected in user code.